helpers_test.go

```diff
@@ -0,0 +1,729 @@
+		eventSeparator()
+	}
+}
+
+type eventCollector struct {
+	streams map[string]*EventStream
+	e       Events
+	mu      sync.Mutex
+	done    chan struct{}
+}
+
+func newCollector() *eventCollector {
+	return &eventCollector{
+		streams: make(map[string]*EventStream),
+		done:    make(chan struct{}),
+		e:       make(Events, 0, 8),
+	}
+}
+
+// stop collecting events and return what we've got.
+func (w *eventCollector) stop() Events {
+	return w.stopWait(time.Second)
+}
+
+func (w *eventCollector) stopWait(waitFor time.Duration) Events {
+	waitForEvents()
+
+	time.Sleep(waitFor)
+
+	for _, es := range w.streams {
+		es.Flush(true)
+		es.Stop()
+	}
+
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return w.e
+}
+
+type Events []Event
+
+func (e Events) String() string {
+	b := new(strings.Builder)
+	for i, ee := range e {
+		if i > 0 {
+			b.WriteString("\n")
+		}
+
+		fmt.Fprintf(b, "%-8v %s", ee.Flags.String(), filepath.ToSlash(ee.Path))
+	}
+	return b.String()
+}
+
+func (e Events) TrimPrefix(prefix string) Events {
+	prefix = strings.TrimPrefix(prefix, "/")
+
+	for i := range e {
+		if e[i].Path == prefix {
+			e[i].Path = "/"
+		} else {
+			e[i].Path = strings.TrimPrefix(e[i].Path, prefix)
+		}
+	}
+	return e
+}
+
+func (e Events) copy() Events {
+	cp := make(Events, len(e))
+	copy(cp, e)
+	return cp
+}
+
+// Create a new Events list from a string; for example:
+//
+//	CREATE        path
+//	CREATE|WRITE  path
+//
+// Every event is one line, and any whitespace between the event and path are
+// ignored. The path can optionally be surrounded in ". Anything after a "#" is
+// ignored.
+//
+// Platform-specific tests can be added after GOOS:
+//
+//	# Tested if nothing else matches
+//	CREATE   path
+//
+//	# Windows-specific test.
+//	windows:
+//	  WRITE    path
+//
+// You can specify multiple platforms with a comma (e.g. "windows, linux:").
+// "kqueue" is a shortcut for all kqueue systems (BSD, macOS).
+func newEvents(t *testing.T, s string) Events {
+	t.Helper()
+
+	var (
+		lines  = strings.Split(s, "\n")
+		groups = []string{""}
+		events = make(map[string]Events)
+	)
+	for no, line := range lines {
+		if i := strings.IndexByte(line, '#'); i > -1 {
+			line = line[:i]
+		}
+		line = strings.TrimSpace(line)
+		if line == "" {
+			continue
+		}
+		if strings.HasSuffix(line, ":") {
+			groups = strings.Split(strings.TrimRight(line, ":"), ",")
+			for i := range groups {
+				groups[i] = strings.TrimSpace(groups[i])
+			}
+			continue
+		}
+
+		fields := strings.Fields(line)
+		if len(fields) != 2 && len(fields) != 4 {
+			if strings.ToLower(fields[0]) == "empty" || strings.ToLower(fields[0]) == "no-events" {
+				for _, g := range groups {
+					events[g] = Events{}
+				}
+				continue
+			}
+			t.Fatalf("newEvents: line %d: needs 2 or 4 fields: %s", no+1, line)
+		}
+
+		var resultFlags EventFlags
+
+		flags := strings.Split(fields[0], "|")
+		for _, f := range flags {
+			resultFlags = resultFlags.set(eventFlagsPossible[f])
```

Comment:
I don't think this is case-insensitive? It should be: "created", "CREATED", and "Created" should all work. No reason to be pedantic about capitals here.